### PR TITLE
cmd/geth: implement vulnerability check

### PR DESCRIPTION
**Pull request by holiman**
Adds a command to do a vulnerability check
No args: uses the current version: 
```
[user@work go-ethereum]$ ./build/bin/geth version-check
INFO [11-16|21:24:42.932] Checking vulnerabilities                 version=Geth/v1.9.25-unstable-99d756f0-20201116/linux-amd64/go1.15.5
```
Args can be used to check some specific version: 
```
[user@work go-ethereum]$ ./build/bin/geth version-check "Geth/v1.9.23-unstable-92c56eb8-20201116/linux-amd64/go1.15.1"
INFO [11-16|21:24:53.628] Checking vulnerabilities                 version=Geth/v1.9.23-unstable-92c56eb8-20201116/linux-amd64/go1.15.1
## Vulnerable to GETH-2020-01 (CorruptedDAG)

Severity: Medium
Summary : Mining nodes will generate erroneous PoW on epochs > `385`.
Fixed in: v1.9.24
References:
	- https://github.com/ethereum/go-ethereum/pull/21793
	- https://blog.ethereum.org/2020/11/12/geth_security_release/
	- https://github.com/ethereum/go-ethereum/commit/567d41d9363706b4b13ce0903804e8acf214af49

## Vulnerable to GETH-2020-02 (GoCrash)

Severity: Critical
Summary : A denial-of-service issue can be used to crash Geth nodes during block processing, due to an underlying bug in Go (CVE-2020-28362) versions < `1.15.5`, or `<1.14.12`
Fixed in: v1.9.24
References:
	- https://blog.ethereum.org/2020/11/12/geth_security_release/
	- https://groups.google.com/g/golang-announce/c/NpBGTTmKzpM
	- https://github.com/golang/go/issues/42552
```
Or lke this: 
```
[user@work go-ethereum]$ ./build/bin/geth version-check "Geth/v1.9.23-unstable-92c56eb8-20201116/linux-amd64/go1.16.1"
INFO [11-16|21:25:06.502] Checking vulnerabilities                 version=Geth/v1.9.23-unstable-92c56eb8-20201116/linux-amd64/go1.16.1
## Vulnerable to GETH-2020-01 (CorruptedDAG)

Severity: Medium
Summary : Mining nodes will generate erroneous PoW on epochs > `385`.
Fixed in: v1.9.24
References:
	- https://github.com/ethereum/go-ethereum/pull/21793
	- https://blog.ethereum.org/2020/11/12/geth_security_release/
	- https://github.com/ethereum/go-ethereum/commit/567d41d9363706b4b13ce0903804e8acf214af49

```

### CODE REVIEW

File: cmd/geth/version_check.go [hunk 1]
```diff
@@ -0,0 +1,95 @@
...
+	"github.com/ethereum/go-ethereum/log"
+	"github.com/ethereum/go-ethereum/params"
+	"gopkg.in/urfave/cli.v1"
+	"io/ioutil"
+	"net/http"
+	"regexp"
+	"runtime"
+)
+
+const versionFeed = "https://geth.ethereum.org/docs/vulnerabilities/vulnerabilities.json"
```
Comment (rjl493456442) [inline cmd/geth/version_check.go:31]: Is there any signature we can verify the JSON we get is correct?

**Comment by holiman** [comment]
Lifting this out ot top level: 
> Is there any signature we can verify the JSON we get is correct?

Good that you raised the point, I've also been considering this. 

Some arguments about why we would _not_ need it: 

- The information is _informational_ only, so the most that an attacker could do is present the user with bogus vulnerability information. 
- The integrity is ensured via traditional CA infrastructure (via cloudflare / github), and git. A modification would have to either break the https integrity or the integrity of the gh-pages branch on github, e.g. by making an unauthorized commit. 
- If geth _fails_ to verify the signature, then it's difficult for the user to know what to do. Update? Not update? 

Some arguments about why it would be good to have it: 
- It is fairly easy to add, and would be  good starting point to explore `signify` (https://github.com/ethereum/go-ethereum/pull/21798) 
- If we have it, we can change the check to use other channels for publishing the data, e.g we could publish it on various platforms across the world to alleviate centralization concerns (github / cloudflare) or phone-home concerns. Then users could e.g. do `geth --version-check -u <local_url>`, or geth could randomize  which out of N platforms do to the check against. 
  - Ability to specify the `url` would also be good for local testing
  - However, having stale security-info will _not_ work to increase user security, so if we do this we should republish signatures every 2 weeks, and consider it stale if it is older than 2 weeks.

**Comment by karalabe** [comment]
> The information is informational only, so the most that an attacker could do is present the user with bogus vulnerability information.

Even your vuln descriptions contain a whole lot of links to visit. I think it would be easy for an attacker to get the user to update Geth from a weird path if they can control the input (after all, we've even given bug bounties for the same thing in the logs). IMHO it's a valid thing to ensure the bug listing is signed.

**Comment by karalabe** [comment]
Also, the domain is controlled by the EF. Do you trust them? :trollface:

**Comment by holiman** [comment]
We should probably have a set of keys, one for the travis builder and a set of developer keys.

**Comment by holiman** [comment]
How to generate keys, @karalabe @fjl :
```
$ minisign -G
Please enter a password to protect the secret key.

Password: 
Password (one more time): 
Deriving a key from the password in order to encrypt the secret key... done

The secret key was saved as /home/user/.minisign/minisign.key - Keep it secret!
The public key was saved as minisign.pub - That one can be public.
```

**Comment by holiman** [comment]
@gballet regarding compatibility with `openbsd-signify`

`openbsd-signify` _can_ be used to verify our `minisig`-generated signatures OK:

```
user@debian-work:~/QubesSync/all/go-ethereum/cmd/geth/testdata/vcheck$ signify-openbsd -V -p minisign.pub -m data.json -x ./sigs/vulnerabilities.json.minisig.1
Signature Verified
user@debian-work:~/QubesSync/all/go-ethereum/cmd/geth/testdata/vcheck$ signify-openbsd -V -p minisign.pub -m data.json -x ./sigs/vulnerabilities.json.minisig.2
Signature Verified
user@debian-work:~/QubesSync/all/go-ethereum/cmd/geth/testdata/vcheck$ signify-openbsd -V -p minisign.pub -m data.json -x ./sigs/vulnerabilities.json.minisig.3
Signature Verified
```

However, creating them fails: 
```
$ signify-openbsd -S -s signify.key -m data.json 
signify-openbsd: invalid base64 encoding in signify.key
```
Which is because the key is stored differently: 
```
user@debian-work:~/QubesSync/all/go-ethereum/cmd/geth/testdata/vcheck$ cat signifykey.sec 
untrusted comment: signify secret key
RWRCSwAAACpLQDLawSQCtI7eAVIvaiHzjTsTyJsfV5aKLNhZb0KdAWeICXJGa93/bHAcsY6jUh9I8RdEcDWEoGxmaXZC+IdVBPxDpkix9fBRGEUdKWHi3dOfqME0YRzErWI5AVg3cRw=
user@debian-work:~/QubesSync/all/go-ethereum/cmd/geth/testdata/vcheck$ cat minisign.key 
untrusted comment: minisign encrypted secret key
RWRTY0Iyz8kmPMKrqk6DCtlO9a33akKiaOQG1aLolqDxs52qvPoAAAACAAAAAAAAAEAAAAAArEiggdvyn6+WzTprirLtgiYQoU+ihz/HyGgjhuF+Pz2ddMduyCO+xjCHeq+vgVVW039fbsI8hW6LRGJZLBKV5/jdxCXAVVQE7qTQ6xpEdO0z8Z731/pV1hlspQXG2PNd16NMtwd9dWw=
```
Which the [minisig docs](https://jedisct1.github.io/minisign/) mention: 

> Signature written by minisign can be verified using OpenBSD's signify tool: public key files and signature files are compatible.
>
> However, minisign uses a slightly different format to store secret keys.

However, it's also the case that `minisign` cannot verify signatures made by `signify`. I've added a test (which fails, so is skipped), and also verified the behaviour on the canonical packages; 
```
[user@work vcheck]$ minisign -V -p ./minisign.pub -m data.json -x ./minisig-sigs/vulnerabilities.json.minisig.1 && echo "OK"
Signature and comment signature verified
Trusted comment: timestamp:1605618622	file:vulnerabilities.json
OK
[user@work vcheck]$ minisign -V -p ./signifykey.pub -m data.json -x ./signify-sigs/data.json.sig && echo "OK"
Trusted comment not present
```

**Comment by holiman** [comment]
So in essense

| signed with | verify with | Status|
|----------------|----------------|--------|
| minisign | minisign | OK|
| minisign | signify | OK|
| signify | minisign | FAIL| 
| signify | signify | OK |

**Comment by holiman** [comment]
This PR depends on https://github.com/ethereum/go-ethereum/pull/21889, but is otherwise ready for merge, IMO. PTAL

**Comment by holiman** [comment]
TODO: add CVE numbers to the json

**Comment by holiman** [comment]
Current status: 
```
[user@work go-ethereum]$ ./build/bin/geth version-check
INFO [11-24|13:30:03.069] Checking vulnerabilities                 version=Geth/v1.9.25-unstable-6bebea97-20201123/linux-amd64/go1.15.5 url=https://geth.ethereum.org/docs/vulnerabilities/vulnerabilities.json
No vulnerabilities found
[user@work go-ethereum]$ ./build/bin/geth version-check --check.version="=Geth/v1.9.25-unstable-6bebea97-20201123/linux-amd64/go1.15.2"
INFO [11-24|13:30:39.736] Checking vulnerabilities                 version="=Geth/v1.9.25-unstable-6bebea97-20201123/linux-amd64/go1.15.2" url=https://geth.ethereum.org/docs/vulnerabilities/vulnerabilities.json
## Vulnerable to GETH-2020-02 (GoCrash)

Severity: Critical
Summary : A denial-of-service issue can be used to crash Geth nodes during block processing, due to an underlying bug in Go (CVE-2020-28362) versions < `1.15.5`, or `<1.14.12`
Fixed in: v1.9.24
References:
        - https://blog.ethereum.org/2020/11/12/geth_security_release/
        - https://groups.google.com/g/golang-announce/c/NpBGTTmKzpM
        - https://github.com/golang/go/issues/42552

[user@work go-ethereum]$ ./build/bin/geth version-check --check.version="=Geth/v1.9.12-unstable-6bebea97-20201123/linux-amd64/go1.15.2"
INFO [11-24|13:30:57.229] Checking vulnerabilities                 version="=Geth/v1.9.12-unstable-6bebea97-20201123/linux-amd64/go1.15.2" url=https://geth.ethereum.org/docs/vulnerabilities/vulnerabilities.json
## Vulnerable to GETH-2020-02 (GoCrash)

Severity: Critical
Summary : A denial-of-service issue can be used to crash Geth nodes during block processing, due to an underlying bug in Go (CVE-2020-28362) versions < `1.15.5`, or `<1.14.12`
Fixed in: v1.9.24
References:
        - https://blog.ethereum.org/2020/11/12/geth_security_release/
        - https://groups.google.com/g/golang-announce/c/NpBGTTmKzpM
        - https://github.com/golang/go/issues/42552

## Vulnerable to GETH-2020-03 (ShallowCopy)

Severity: Critical
Summary : A consensus flaw in Geth, related to `datacopy` precompile
Fixed in: v1.9.17
References:
        - https://blog.ethereum.org/2020/11/12/geth_security_release/

```

Afaik this PR "done"

**Comment by holiman** [comment]
> Afaik this PR "done"

Ok it wasn't quite, but now :)

File: cmd/geth/version_check.go [hunk 2]
```diff
@@ -0,0 +1,155 @@
...
+	if err = verifySignature(gethPubKeys, data, sig); err != nil {
+		return err
+	}
+	var vulns []vulnJson
+	if err = json.Unmarshal(data, &vulns); err != nil {
+		return err
+	}
+	allOk := true
+	for _, vuln := range vulns {
+		r, err := regexp.Compile(vuln.Check)
```
Comment (fjl) [inline cmd/geth/version_check.go:83]: regexp.CompilePOSIX

**Comment by holiman** [comment]
TODOs outside of this PR: 

- Fix the error in the regexp check for mining issue, `1.9.12` etc, 
- Add CVE numbers to the vuln json

**Review by fjl — APPROVED** [review]
Code is fine, but we found an issue in the 'production' vulnerabilities.json. The regular expression for the CorruptedDAG issue does not cover v1.9.0, for example. We decided to fix this by extending the version matcher to also use "introduced" and "fixed" JSON fields.

**Review by kaycoinsUSA — COMMENTED** [review]
Good

**Review by kaycoinsUSA — COMMENTED** [review]
Good

**Review by KaycoinsInc — APPROVED** [review]

**Review by chaddychad47 — COMMENTED** [review]
Sync

**Review by totongcute — COMMENTED** [review]
sub

**Review by totongcute — COMMENTED** [review]
s

**Comment by zhiqiangxu** [comment]
> > Afaik this PR "done"
> 
> Ok it wasn't quite, but now :)

why `dataCopy` doesn't do the copy, but returns in bytes directly?

**Pull request merged**